runtime/test/reference-test.js

```diff
@@ -173,4 +173,182 @@ describe('references', function() {
     const refStore = arc._stores[1];
     await assertSingletonWillChangeTo(arc, refStore, 'storageKey', arc._storageProviderFactory.baseStorageKey(Type.newEntity(manifest.schemas.Result), 'in-memory'));
   });
+
+  it('can deal with references in schemas', async () => {
+    let loader = new StubLoader({
+      manifest: `
+        schema Result
+          Text value
+        
+        particle ExtractReference in 'extractReference.js'
+          in Foo {Reference<Result> result} referenceIn
+          out Result rawOut
+          
+        recipe
+          create 'input:1' as handle0
+          create 'output:1' as handle1
+          ExtractReference
+            referenceIn <- handle0
+            rawOut -> handle1
+        `,
+      'extractReference.js': `
+        defineParticle(({Particle}) => {
+          return class Dereferencer extends Particle {
+            setHandles(handles) {
+              this.output = handles.get('rawOut');
+            }
+
+            async onHandleSync(handle, model) {
+              console.log('sync', handle.name, model);
+            }
+
+            async onHandleUpdate(handle, update) {
+              if (handle.name == 'referenceIn') {
+                let result = update.data.result;
+                await result.dereference();
+                this.output.set(result.entity);
+              }
+            }
+
+            onHandleDesync(handle) {
+            }
+          }
+        });
+      `
+    });
+
+    let pecFactory = function(id) {
+      let channel = new MessageChannel();
+      new ParticleExecutionContext(channel.port1, `${id}:inner`, loader);
+      return channel.port2;
+    };
+    let arc = new Arc({id: 'test:0', pecFactory, loader});
+
+    let manifest = await Manifest.load('manifest', loader);
+    let recipe = manifest.recipes[0];    
+    assert.isTrue(recipe.normalize());
+    assert.isTrue(recipe.isResolved());
+    await arc.instantiate(recipe);
+
+    const inMemoryEngine = arc._storageProviderFactory._storageInstances['in-memory'];
+    const baseStoreType = Type.newEntity(manifest.schemas.Result);
+    const backingStore = await inMemoryEngine.baseStorageFor(baseStoreType, inMemoryEngine.baseStorageKey(baseStoreType));
+    await backingStore.store({id: 'id:1', rawData: {value: 'what a result!'}}, ['totes a key']);
+    
+    const refStore = arc._stores[1];
+    assert.equal(refStore._type.entitySchema.name, 'Foo');
+    await refStore.set({id: 'id:2', rawData: {result: {id: 'id:1', storageKey: backingStore.storageKey}}});
+
+    await assertSingletonWillChangeTo(arc, arc._stores[0], 'value', 'what a result!');
+  });
+
+  it('can construct references in schemas', async () => {
+    let loader = new StubLoader({
+      manifest: `
+        schema Result
+          Text value
+        
+        particle Referencer in 'referencer.js'
+          in [Result] inResult
```

Comment:
FWIW it took me some time to understand what is happening here. Maybe if you feel like it we could refactor for readability?

E.g.
```
schema Book
  Number isbn
  Text title
  Text content

particle BookLookup in 'book-lookup.js'
  in [Book] library
  inout [Lookup {Number isbn, Reference<Book> result}] isbnLookup
```

This may not do what you want though. It's still somewhat unclear to me why do you need both `inFoo` and `out`, and why the test case stores entities in both - maybe you can explain offline?

Reply:
I'll add a big comment instead; I don't really want to tie this down to a "real" example because it's a pretty synthetic test.